nbgrader/apps/zipcollectapp.py

```diff
@@ -0,0 +1,486 @@
+        src_files: list
+            List of all files in the `extracted_directory`
+
+        Returns:
+        --------
+        Dict: Collected data object of the form
+            {
+                student_id: {
+                    src_files: [src_file1, ...],
+                    dest_files: [dest_file1, ...],
+                    file_ids: [file_id1, ...],
+                    timestamps: [timestamp1, ...],
+                }, ...
+            }
+        """
+        self.log.info("Start collecting files...")
+        released_path = self._format_path(
+            self.release_directory, '.', self.assignment_id)
+        released_notebooks = find_all_notebooks(released_path)
+        if not released_notebooks:
+            self.log.warn(
+                "No release notebooks found for assignment {}, did you forget "
+                "to run 'nbgrader assign'?".format(self.assignment_id)
+            )
+
+        data = dict()
+        invalid_files = 0
+        processed_files = 0
+        for _file in src_files:
+            self.log.info("Processing file: {}".format(_file))
+            info = self.collector_plugin_inst.collect(_file)
+            if not info or info is None:
+                self.log.warn("Skipped. No match information provided.")
+                invalid_files += 1
+                continue
+
+            for key in ['student_id', 'file_id']:
+                if key not in info.keys():
+                    self.fail(
+                        "Expected processor {} to provide the {} from the "
+                        "submission file name."
+                        "".format(self.collector_plugin.__name__, key)
+                    )
+                if not isinstance(info[key], six.string_types):
+                    self.fail(
+                        "Expected processor {} to provide a string for {} from "
+                        "the submission file name."
+                        "".format(self.collector_plugin.__name__, key)
+                    )
+
+            root, ext = os.path.splitext(_file)
+            file_id = os.path.splitext(os.path.basename(info['file_id']))[0]
+            submission = '{}{}'.format(file_id, ext)
+            if ext in ['.ipynb'] and submission not in released_notebooks:
+                if self.strict:
+                    self.log.warn("Skipped. Invalid notebook name.")
+                    invalid_files += 1
+                    continue
+                self.log.warn("Invalid notebook name.")
+
+            if self._find_student(info) is None:
+                self.log.warn(
+                    "Skipped. Student {} not found in gradebook."
+                    "".format(info['student_id'])
+                )
+                invalid_files += 1
+                continue
+
+            submitted_path = self._format_path(
+                self.submitted_directory, info['student_id'], self.assignment_id)
+            dest_path = os.path.join(submitted_path, submission)
```

Comment:
It would be nice if this is something that could be returned in `info` so it could be configured by a plugin.

Also, if I have subdirectories in my assignments, then this will cause all files to be moved out of the subdirectories into a flat directory, which is undesirable. I think the default behavior should be to preserve directory structure.